src/runner.py

```diff
@@ -88,596 +47,416 @@ class LxdInstanceConfig(TypedDict):
+                raise RunnerRemoveError(f"Unable to stop LXD instance for runner {self.name}")
 
+        logger.info("Removing LXD instance for runner: %s", self.name)
 
-class RunnerError(Exception):
-    """Generic runner error."""
+        if isinstance(self.path, GitHubRepo):
+            self._github.actions.delete_self_hosted_runner_from_repo(
+                owner=self.path.owner, repo=self.path.repo, runner_id=self.name
+            )
+        elif isinstance(self.path, GitHubOrg):
+            self._github.actions.delete_self_hosted_runner_from_org(
+                org=self.path.org, runner_id=self.name
+            )
 
+        if self.instance is None:
+            return
 
-class RunnerCreateError(RunnerError):
-    """Error for runner creation failure."""
+        if self.instance.status == "Running":
+            self.instance.stop(wait=True)
+            time.sleep(30)
 
+            check_shutdown()
 
-class RunnerRemoveError(RunnerError):
-    """Error for runner removal failure."""
+            try:
+                self.instance.delete(wait=True)
+            except Exception:  # nosec B110
+                # Ephemeral containers should auto-delete when stopped;
+                # this is just a fall-back.
+                pass
+        else:
+            # We somehow have a non-running instance which should have been
+            # ephemeral. Try to delete it and allow any errors doing so to
+            # surface.
+            try:
+                self.instance.delete(wait=True)
+            except Exception as err:
+                raise RunnerRemoveError() from err
 
+    @retry(tries=10, delay=1, logger=logger)
+    def _create_instance(
+        self, image: str, resources: VirtualMachineResources, ephemeral: bool = True
+    ) -> pylxd.models.Instance:
+        """Create a instance of runner.
 
-class RunnerStartError(RunnerError):
-    """Error for runner start failure."""
+        Args:
+            image: Image to launch the instance hosting the runner.
+            resources: Configuration of the virtual machine resources.
+            ephemeral: Whether the instance is ephemeral. Defaults to True.
 
+        Returns:
+            LXD instance of the runner.
+        """
+        logger.info("Creating an LXD instance for runner: %s", self.name)
 
-class RunnerManager:
-    """Class for querying and controlling the runners.
+        self._ensure_runner_profile()
+        resource_profile = self._get_resource_profile(resources)
 
-    Attributes:
-        proxies (Dict[str, str]): Mapping of proxy env vars.
-        path (str): GitHub repository path in the format '<org>/<repo>', or the GitHub org name.
-        app_name (str): An name for the set of runners.
-        reconcile_interval (int): Number of minutes between each reconciliation of runners state.
-    """
+        # Create runner instance.
+        instance_config: LxdInstanceConfig = {
+            "name": self.name,
```

Comment:
No risk of instance already existing ?

Reply:
The instance name will be generated with UUID v4, so this should not be an issue.